stanford_soe_r25.module

```diff
@@ -0,0 +1,889 @@
+      if (!empty($rooms[$room_id]['approver_secgroup_id'])) {
+        // get an array of email addresses for the people in the room's approver security group
+        $approvers = _stanford_soe_r25_security_group_emails($rooms[$room_id]['approver_secgroup_id']);
+        // see if the Drupal user's email address is in the array of approvers
+        $can_cancel = in_array($user->mail, $approvers);
+      }
+    }
+
+    if ($can_cancel) {
+      // if the user can cancel (or confirm) the event, return the event's XML arrays to the caller
+      $output = $result;
+    }
+    else {
+      // otherwise just output false
+      $output = FALSE;
+    }
+  }
+  else {
+    // set an error message if we couldn't contact 25Live.
+    drupal_set_message('Unable to retrieve data from 25Live. Please try again later.', 'error', FALSE);
+    $output = FALSE;
+  }
+  return $output;
+}
+
+// this function returns an array of email addresses for a 25Live security group
+// whose id is stored with a room configuration. The array will be stored in the
+// Drupal cache table unless forced to refresh with the $reset paramater. Saving
+// the room configuration in its admin page will reset this cached data.
+function _stanford_soe_r25_security_group_emails($secgroup_id = NULL, $reset = FALSE) {
+  if (empty($secgroup_id)) {
+    return array();
+  }
+
+  // if the information is in the cache and we're not resetting it
+  // then return it from the cache
+  $cache_id = 'stanford_soe_r25:approvers:' . $secgroup_id;
+  if (!$reset) {
+    $list = &drupal_static(__FUNCTION__);
+    if (!$list) {
+      $cache = cache_get($cache_id);
+      if ($cache && !empty($cache->data)) {
+        $list = $cache->data;
+        return $list;
+      }
+    }
+  }
+
+  // otherwise, request the contact ids and email addresses for the security
+  // group members from the 25Live API and store them in the cache.
+  $list = array();
+  $sec_result = _stanford_soe_r25_api_call('r25users', NULL, NULL, $secgroup_id);
+  if (!empty($sec_result['index']['R25:CONTACT_ID']) && is_array($sec_result['index']['R25:CONTACT_ID'])) {
+    foreach ($sec_result['index']['R25:CONTACT_ID'] as $key => $value) {
+      if (!empty($sec_result['vals'][$value]['value'])) {
+        $email = '';
+        if (!empty($sec_result['index']['R25:PRIMARY_EMAIL'][$key]) &&
+          !empty($sec_result['vals'][$sec_result['index']['R25:PRIMARY_EMAIL'][$key]]['value'])
+        ) {
+          $email = $sec_result['vals'][$sec_result['index']['R25:PRIMARY_EMAIL'][$key]]['value'];
+        }
+        $list[$sec_result['vals'][$value]['value']] = $email;
+      }
+    }
+  }
+  cache_set($cache_id, $list);
+  return $list;
+}
+
+// This is where we hook into the reservation form to change things for
+// insidesoe field placement and overrides.
+function stanford_soe_r25_form_stanford_soe_r25_reservation_alter(&$form, &$form_state, $form_id) {
```

Comment:
why is it necessary to hook into our own code?

Reply:
Good pick up.

Reply:
Also missing function doc block.